src/ws/cert-session.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This program opens a session for a given user and runs the bridge in
+ * it, using TLS client certificate authentication and mapping it to a user
+ * through sssd (https://www.freeipa.org/page/V4/User_Certificates).
+ */
+
+#include "config.h"
+#include <stdlib.h>
+
+#include <gio/gio.h>
```

Comment:
I don't think using GLib in a setuid process is sensible from a security perspective, long term. GLib does unpredictable things in process initialization, global variables, threading, child reaping, and other activities. In our final solution this should not be present.

This commit however says "unsafe" so I'm fine with this for the time being.

Reply:
Agreed. I'd really like to use sd-dbus for that, but that doesn't yet exist on RHEL 7. I need to check with the stakeholder if having that in RHEL 8 is suffiicent. If so, then sd-dbus is IMHO the least intrusive choice, otherwise we'd have to grow a dependency to libdbus.

Reply:
Lets discuss a solution where we do this upfront while accepting the certificate and avoid DBus in a setuid process. Not set on it but it's a possible solution.

Reply:
Note that this D-Bus API is restricted to root. Also, this is a crucial security-sensitive mapping, so it can't be done as cockpit-ws or the target user. This must be in a small trusted helper, so a suid process is necessary. Of course it could suid to a different system user and the sssd policy be extended to also allow that (just more moving parts). But what's bad about doing this as root? If there's a vulnerability in the D-Bus protocol that allows the server side to run arbitrary code in the caller, then we are screwed anyway (there's plenty of D-Bus clients running as root during boot even).

Reply:
FTR, no need to backport this to RHEL 7, so we can use sd-dbus if we want to.